vmm/task/src/youki.rs

```diff
@@ -0,0 +1,646 @@
+            let (pio, container_io) = create_io(&p.id, self.io_uid, self.io_gid, &p.stdio)?;
+            //exec_opts.io = pio.io.as_ref().cloned();
+            (None, Some(pio), Some(container_io))
+        };
+
+        let probe_path = format!("{}/{}-process.json", self.bundle, &p.id);
+        let spec_str = serde_json::to_string(&self.spec)
+            .map_err(other_error!(e, "failed to marshall exec spec to string"))?;
+        tokio::fs::write(&probe_path, &spec_str)
+            .await
+            .map_err(other_error!(e, "failed to write spec to process.json"))?;
+
+        let container_id = self.container_id.clone();
+        let socket_path = socket.as_ref().map(|p| p.path.clone());
+        let probe_path_clone = probe_path.clone();
+        let exec_result = spawn_blocking(move || {
+            let builder = ContainerBuilder::new(container_id, SyscallType::default())
+                .with_root_path(PathBuf::from(YOUKI_DIR))
+                .map_err(other_error!(e, "failed to set youki root path"))?
+                .with_console_socket(socket_path);
+
+            // if let Some(f) = container_io {
+            //     if let Some(fd) = f.stdin {
+            //         builder = builder.with_stdin(fd);
+            //     }
+            //     if let Some(fd) = f.stdout {
+            //         builder = builder.with_stdout(fd);
+            //     }
+            //     if let Some(fd) = f.stderr {
+            //         builder = builder.with_stderr(fd);
+            //     }
+            // }
+
+            builder
+                .as_tenant()
+                .with_detach(true)
+                .with_process(Some(&probe_path_clone))
+                .build()
+                .map_err(other_error!(
+                    e,
+                    "failed to exec process in youki container "
+                ))
+        })
+        .await
+        .map_err(other_error!(e, "failed to wait exec thread "))?;
+        tokio::fs::remove_file(&probe_path)
+            .await
+            .unwrap_or_default();
+        match exec_result {
+            Ok(pid) => {
+                copy_io_or_console(p, socket, pio, p.lifecycle.exit_signal.clone()).await?;
+                p.pid = pid.as_raw();
+                p.state = Status::RUNNING;
+            }
+            Err(e) => {
+                if let Some(s) = socket {
+                    s.clean().await;
+                }
+                return Err(other!("failed to start youki exec: {}", e));
+            }
+        }
+        Ok(())
+    }
+
+    async fn kill(
+        &self,
+        p: &mut ExecProcess,
+        signal: u32,
+        _all: bool,
+    ) -> containerd_shim::Result<()> {
+        if p.pid <= 0 {
+            Err(Error::FailedPreconditionError(
+                "process not created".to_string(),
+            ))
+        } else if p.exited_at.is_some() {
+            Err(Error::NotFoundError("process already finished".to_string()))
+        } else {
```

Comment:
The most possiable branch of this `if-else` is in the last condition, how about put it in the first condition?

Reply:
I think this is also the same with container.rs, Why should we have to put the most possible branch in the first condition?
I think the original code should be like this:
```
if p.pid <= 0 {
    return Err(Error::FailedPreconditionError(
        "process not created".to_string(),
    ))
} 
if p.exited_at.is_some() {
    return Err(Error::NotFoundError("process already finished".to_string()))
}

kill(
    Pid::from_raw(p.pid),
    nix::sys::signal::Signal::try_from(signal as i32).unwrap(),
)
.map_err(Into::into)
```
it maybe changed into the what is like now, because in the rust check tool the return is not welcome.